packages/vm/src/calls.rs

```diff
@@ -734,6 +738,93 @@ mod tests {
         assert_eq!(query_response.as_slice(), b"{\"verifier\":\"verifies\"}");
     }
 
+    #[test]
+    fn float_instrs_are_deterministic() {
+        #[derive(Debug, serde::Serialize, serde::Deserialize)]
+        #[serde(rename_all = "snake_case")]
+        pub enum Value {
+            U32(u32),
+            U64(u64),
+            F32(u32),
+            F64(u64),
+        }
+
+        let mut instance = mock_instance_with_options(
+            FLOATY2,
+            MockInstanceOptions {
+                gas_limit: u64::MAX,
+                memory_limit: None,
+                ..Default::default()
+            },
+        );
+
+        // init
+        let info = mock_info("creator", &[]);
+        call_instantiate::<_, _, _, Empty>(&mut instance, &mock_env(), &info, br#"{}"#)
+            .unwrap()
+            .unwrap();
+
+        // query instructions
+        let msg = br#"{"instructions":{}}"#;
+        let contract_result = call_query(&mut instance, &mock_env(), msg)
+            .unwrap()
+            .unwrap();
+        let instructions: Vec<String> = from_slice(&contract_result).unwrap();
+        // little sanity check
+        assert_eq!(instructions.len(), 70);
+
+        const RUNS_PER_INSTRUCTION: u64 = 150;
+        let mut hasher = Sha256::new();
+        for instr in &instructions {
+            for seed in 0..RUNS_PER_INSTRUCTION {
+                // query some input values for the instruction
+                let args: Vec<Value> = from_slice(
+                    &call_query(
+                        &mut instance,
+                        &mock_env(),
+                        format!(
+                            r#"{{"random_args_for":{{ "instruction": "{instr}", "seed": {seed}}}}}"#
+                        )
+                        .as_bytes(),
+                    )
+                    .unwrap()
+                    .unwrap(),
+                )
+                .unwrap();
+
+                // build the run message
+                let arg_str = args
+                    .iter()
+                    .map(|a| {
+                        let serialized = to_vec(a).unwrap();
+                        core::str::from_utf8(&serialized).unwrap().to_string()
+                    })
+                    .collect::<Vec<_>>()
+                    .join(",");
```

Comment:
Here you do manual JSON creation in two different places: 1. join with "," and 2. wrap in "[" / "]" below. I think you can just convert the Vec<String> to JSON in one go.